backend/src/process.py

```diff
@@ -264,6 +264,8 @@ def __init__(
             ProgressController() if not parent_executor else parent_executor.progress
         )
 
+        self.completed_node_ids = []
```

Comment:
Please make this a set. You only use `completed_node_ids` to add ids not already in `completed_node_ids` and to get its length. Both are O(1) operations for sets.